crates/stark-backend/src/prover/hal.rs

```diff
@@ -0,0 +1,180 @@
+    /// quotient polynomial evaluation can be done on the LDE domain. This avoids a separate
+    /// cosetDFT step.
+    ///
+    /// For each RAP, `quotient_degree` is the number of quotient chunks that were committed.
+    /// So `quotient_degree = quotient_domain_size / trace_domain_size`.
+    ///
+    /// The `constraints` contains the serializable symbolic constraints of each RAP across all challenge phases.
+    ///
+    /// Quotient polynomials for multiple RAP matrices are committed together into a single commitment.
+    /// The quotient polynomials can be committed together even if the corresponding trace matrices
+    /// are committed separately.
+    fn eval_and_commit_quotient(
+        &self,
+        challenger: &mut PB::Challenger,
+        constraints: &[&SymbolicExpressionDag<PB::Val>],
+        extended_views: Vec<RapView<PB::MatrixView, PB::Val, PB::Challenge>>,
+        quotient_degrees: &[u8],
+    ) -> (PB::Commitment, PB::PcsDataView);
+}
+
+/// Polynomial commitment scheme (PCS) opening proof generator.
+pub trait OpeningProver<PB: ProverBackend> {
+    /// Opening proof for multiple RAP matrices, where
+    /// - (for now) each preprocessed trace matrix has a separate commitment
+    /// - main trace matrices can have multiple commitments
+    /// - for each after_challenge phase, all matrices in the phase share a commitment
+    /// - quotient poly chunks are all committed together
+    fn open(
+        &self,
+        challenger: &mut PB::Challenger,
+        // For each preprocessed trace commitment, the prover data and
+        // the log height of the matrix, in order
+        preprocessed: Vec<PB::PcsDataView>,
+        // For each main trace commitment, the prover data and
+        // the log height of each matrix, in order
+        // Note: this is all one challenge phase.
+        main: Vec<PB::PcsDataView>,
+        // `after_phase[i]` has shared commitment prover data for all matrices in phase `i + 1`.
+        after_phase: Vec<PB::PcsDataView>,
+        // Quotient poly commitment prover data
+        quotient_data: PB::PcsDataView,
+        // Quotient degree for each RAP committed in quotient_data, in order
+        quotient_degrees: &[u8],
+    ) -> PB::OpeningProof;
+}
+
+/// Trait to manage data transport of prover types from host to device.
+pub trait DeviceDataAdapter<SC, PB>
+where
+    SC: StarkGenericConfig,
+    PB: ProverBackend<Val = Val<SC>, Challenge = SC::Challenge, Commitment = Com<SC>>,
+{
+    /// Transport the proving key to the device, filtering for only the provided `air_ids`.
+    fn transport_pk_to_device<'a>(
+        &'a self,
+        mpk: &'a MultiStarkProvingKey<SC>,
+        air_ids: Vec<usize>,
+    ) -> MultiStarkProvingKeyView<'a, PB, &'a RapPhaseSeqProvingKey<SC>>
+    where
+        SC: 'a;
+
+    fn transport_matrix_to_device(&self, matrix: &Arc<RowMajorMatrix<Val<SC>>>) -> PB::MatrixView;
+
```

Comment:
We may also want to have a function to transfer matrix from device to host in the case that we want to have a hybrid prover.

Reply:
But we need to be careful as the side effect of this function implies that the gpu memory occupied by the matrix should be deallocated. This means that we need to deactivate all other `PB::MatrixView`s.

Reply:
This function should work, but you have to collect all `MatrixView` to one `Matrix`, and if one of them is missing, then it might cause GPU OOM issue.

Reply:
If you define associated types like `Matrix` or `PcsData` and when it's out of scope (e.g. call the function `fn transport_matrix_to_host(&self, matrix: PB::Matrix) -> RowMajorMatrix` then the rust compiler can make sure there is no dangling pointers / references.